src/db/query-validation.ts

```diff
@@ -0,0 +1,178 @@
+						throw new Error(
+							fieldTypeError(whereNode.key, FieldType.String, headerType)
+						);
+					}
+					break;
+				default:
+					throw new Error(`Unsupported type for key: ${whereNode.key}`);
+			}
+		}
+	}
+}
+
+/**
+ * validateOrderBy validates the 'orderBy' clause of the query.
+ * Currently supports strictly one 'orderBy' condition that must match the 'where' clause key.
+ *
+ * @param {OrderBy<T>[] | undefined} orderBy - The 'orderBy' clause to validate.
+ * @param {string} whereKey - The key used in the 'where' clause for compatibility.
+ * @throws {Error} Throws an error if the 'orderBy' clause is invalid or doesn't match the 'where' clause key.
+ */
+function validateOrderBy<T extends Schema>(
+	orderBy: OrderBy<T>[] | undefined,
+	whereKey: string
+): void {
+	if (orderBy) {
+		if (!Array.isArray(orderBy) || orderBy.length === 0) {
+			throw new Error("Invalid 'orderBy' clause.");
+		}
+
+		// Note: currently we only support one orderBy and it must be the where clause. When we add composite indexes and complex querying, refactor.
+		const orderByObj = orderBy[0];
+
+		if (!["ASC", "DESC"].includes(orderByObj.direction)) {
+			throw new Error("Invalid direction in `orderBy`.");
+		}
+
+		if (orderByObj.key !== whereKey) {
+			throw new Error("'key' in `orderBy` must match `key` in `where` clause");
+		}
+	}
+}
+
+/**
+ * validateSelect validates the 'select' fields of a query.
+ *
+ * @param {SelectField<T>[] | undefined} select - The 'select' clause to validate.
+ * @param {Header[]} headers - List of headers to check for field existence.
+ * @throws {Error} Throws an error if any field in the 'select' clause doesn't exist in headers.
+ */
+function validateSelect<T extends Schema>(
+	select: SelectField<T>[] | undefined,
+	headers: Header[]
+): void {
+	if (select) {
+		if (!Array.isArray(select) || select.length === 0) {
+			throw new Error("Invalid 'selectFields' clause");
+		}
+
+		for (const field of select) {
+			const header = headers.find((h) => h.fieldName === field);
+
+			if (!header) {
+				throw new Error(
+					`'key': ${field as string} in 'selectFields' clause does not exist in dataset.`
+				);
+			}
+		}
+	}
+}
+
+/**
+ * validateQuery checks the structure and syntax of the query against the provided headers.
+ * It ensures that the fields specified in the query are valid and exist in the headers.
+ * If any part of the query is invalid (e.g., a field doesn't exist), it throws an error.
+ *
+ * @param {Query<T>} query - The query object to validate.
+ * @param {Header[]} headers - The headers against which to validate the query fields.
+ * @throws {Error} Throws an error if query is invalid.
+ */
+export async function validateQuery<T extends Schema>(
+	query: Query<T>,
+	headers: Header[]
+): Promise<void> {
+	validateWhere(query.where, headers);
+	validateOrderBy(query.orderBy, query.where![0].key as string);
+	validateSelect(query.select, headers);
```

Comment:
should be good to go! Splitting into helper functions will be useful when we start chaining commands when we query

Reply:
Very clean very nice.